### DISCUSSION TITLE
Script: Restore a node hidden state after a Controller.select( this node )

### DISCUSSION THREAD

**Discussion by lilive** [freeplane/freeplane#1071] [Help]
Hi,

In a mind map view, a node N may be hidden for 3 reasons (as far as I know):
- N is inside a folded branch
- N is filtered
- N is "out of the view", because of a previous "Jump In" command

A script can select this node, and it becomes visible again, regardless of the reason it was hidden:

```groovy
c.select( N )
```

When we do that, the branch is unfolded until N is reached, and N is not filtered or "out of the view" anymore. Very convenient.

My problem is that after this selection, I want to select another node:

```groovy
c.select( N )
// do stuff ...
c.select( anotherNode )
```

This is OK, but I also want to restore the hidden state of N and its ancestors. I don't see how to do this with the current scripting API.
I can save the folding state of the branch and restore it. But I don't know how to restore the filter and the "Jump In" state.
Do I miss something ?

### Some background

I need this because my [Jumper addon](https://github.com/lilive/Freeplane-Jumper) allows the user to search for nodes that match some criterions, and to select one. The addon dialog list the nodes that match, then the user can browse these results. This means that when the user select a node in the result list, without leaving the addon dialog, this node is displayed centered in the view:

![example 1](https://user-images.githubusercontent.com/2719976/229167327-f603e878-bee2-4af8-b95e-7112f5c0d63b.png)

The user can review any node in the result list before to actually select one of them, and each time he see the corresponding node centered in the view:

![example 2](https://user-images.githubusercontent.com/2719976/229168049-b768c93c-7694-4ebd-bdca-c881bee27a04.png)

 To achieve this, I simply do
```groovy
c.select( N )
c.centerOnNode( N )
```

But I would like to preserve the map folding, filtering, and "Jump In" states. I don't want the addon to totally scramble the view state, as the user browse the map to find the node he's looking for. 

At the moment, the addon only preserve the folding state, and can ignore the filtered and the "out of view" nodes. But if the user choose not to ignore them, the addon reveal the filtered nodes and the "out of view" nodes, as they are selected in the dialog.

**Comment by lilive** [comment]
Note:
I use `c.select( N )` to reveal and highlight the node N, but I do not need to actually select it. So, another way to reveal and highlight a node could be OK too (and in fact could be better, because a drawback of the select method is that it also add the node to the visited nodes history).

**Comment by EdoFro** [comment]
two ideas to test:

1. for filtered nodes: execute command "ReapplyFilterAction"
2. maybe you don't need to select it using your script. maybe you can call the freeplane find function using its node's id (or other filter that is fast) and apply command "Find Next" ("QuickFindAction.FORWARD"). this way Freeplane uses its existing routines to show and hide the node

**Reply by lilive** [reply]
Thank you Edo

>     1. for filtered nodes: execute command "ReapplyFilterAction"

It seems to work to restore filtered state. But it do not work to restore "Jump In" state.
I don't know exactly what this command do, so I'm a little bit cautious about using it in this case. Maybe this is a little bit overkill to restore the entire map filter when the script have only touched a part of the map.
 
>     maybe you can call the freeplane find function using its node's id (or other filter that is fast) and apply command "Find Next" ("QuickFindAction.FORWARD")

I do not understand how to use the Freeplane find function within a script. I can do `c.find{ it.id == N.id }` but this not set the Freeplane find pattern, so the command "Find Next" do not take it.

**Comment by sifran-github** [comment]
other than what you mentioned, there is another reason a node can be hidden: it can be a hidden node :)
i mean: view->hidden node and view->show hidden nodes.

i never use this option but it exist.

**Reply by lilive** [reply]
:laughing: 
Indeed !
I never use it myself too.

If a node N is hidden this way, its _visible_ property is false when _View > Show hidden nodes_ is disabled, and true otherwise.
1 - The [API documentation for this property](https://docs.freeplane.org/api/org/freeplane/api/NodeRO.html#isVisible()) do not mention it.
2 - No way to know that N is hidden if _View > Show hidden nodes_ is disabled ?
3 - `c.select( N )` do not reveal the node, unlike it does for filtered and "out of view" nodes.
4 - I don't see how to hide or reveal such a node with the API. Of course we can select it and call the menu entry _View > Hidden node_ that toggle its visibility, but because of the point 2 we are blind about the current hidden state of the node.

This complicate things !

**Comment by macmarrum** [comment]
I use [`c.viewRoot`](https://docs.freeplane.org/api/org/freeplane/api/ControllerRO.html#getViewRoot()) to check "jump-in/-out" state:
```groovy
def viewRoot = c.viewRoot
def isJumpedIn = viewRoot != node.mindMap.root
```

To restore jump-in, I'd go with
```groovy
c.select(viewRoot)
menuUtils.executeMenuItems(['JumpInAction'])
```

**Comment by dpolivaev** [comment]
If you miss any feature in API and find it in the source code, create a pull request extending the API. I think it is the easiest and best way to get what you need and to avoid using internals that might break your scripts in the future.
